// @vant/auto-import-resolver/dist/index.esm.mjs
// src/index.ts
function kebabCase(key) {
  const result = key.replace(/([A-Z])/g, " $1").trim();
  return result.split(" ").join("-").toLowerCase();
}
function getModuleType(options) {
  const { ssr, module = "esm" } = options;
  if (ssr !== void 0) {
    return ssr ? "lib" : "es";
  }
  return module === "cjs" ? "lib" : "es";
}
function getSideEffects(dirName, options) {
  const { importStyle = true } = options;
  if (!importStyle) {
    return;
  }
  const moduleType = getModuleType(options);
  if (importStyle === "less")
    return `vant/${moduleType}/${dirName}/style/less`;
  return `vant/${moduleType}/${dirName}/style/index`;
}
function getAPIMap() {
  const apiMap = /* @__PURE__ */ new Map();
  const api = {
    dialog: [
      "showDialog",
      "closeDialog",
      "showConfirmDialog",
      "setDialogDefaultOptions",
      "resetDialogDefaultOptions"
    ],
    imagePreview: ["showImagePreview"],
    notify: [
      "showNotify",
      "closeNotify",
      "setNotifyDefaultOptions",
      "resetNotifyDefaultOptions"
    ],
    toast: [
      "showToast",
      "closeToast",
      "showFailToast",
      "showLoadingToast",
      "showSuccessToast",
      "allowMultipleToast",
      "setToastDefaultOptions",
      "resetToastDefaultOptions"
    ]
  };
  Object.entries(api).forEach(([importName, apiList]) => {
    apiList.forEach((api2) => {
      apiMap.set(api2, importName);
    });
  });
  return apiMap;
}
function VantResolver(options = {}) {
  const moduleType = getModuleType(options);
  const apiMap = getAPIMap();
  return {
    type: "component",
    resolve: (name) => {
      if (name.startsWith("Van")) {
        const partialName = name.slice(3);
        return {
          name: partialName,
          from: `vant/${moduleType}`,
          sideEffects: getSideEffects(kebabCase(partialName), options)
        };
      }
      if (apiMap.has(name)) {
        const partialName = apiMap.get(name);
        return {
          name,
          from: `vant/${moduleType}`,
          sideEffects: getSideEffects(kebabCase(partialName), options)
        };
      }
    }
  };
}
function VantImports(options = {}) {
  const moduleType = getModuleType(options);
  return {
    [`vant/${moduleType}`]: [...getAPIMap().keys()]
  };
}
export {
  VantImports,
  VantResolver
};
